src/Numerical/LinearAlgebra/Determinant.hpp

```diff
@@ -0,0 +1,201 @@
+               tensor.template get<2, 0>()[s] *
+                   tensor.template get<1, 2>()[s]) +
+          tensor.template get<0, 2>()[s] *
+              (tensor.template get<1, 0>()[s] * tensor.template get<2, 1>()[s] -
+               tensor.template get<2, 0>()[s] * tensor.template get<1, 1>()[s]);
+    }
+    return det;
+  }
+};
+
+template <typename Index>
+struct DeterminantImpl<Symmetry<2, 1>, Index,
+                       std::enable_if_t<Index::dim == 4>> {
+  template <typename T>
+  static auto apply(const T& tensor) {
+    DataVector det(tensor.template get<0, 0>().size());
+    for (size_t s = 0; s < det.size(); s++) {
+      const double tmp1 =
+          tensor.template get<2, 2>()[s] * tensor.template get<3, 3>()[s] -
+          tensor.template get<3, 2>()[s] * tensor.template get<2, 3>()[s];
+      const double tmp2 =
+          tensor.template get<2, 1>()[s] * tensor.template get<3, 3>()[s] -
+          tensor.template get<3, 1>()[s] * tensor.template get<2, 3>()[s];
+      const double tmp3 =
+          tensor.template get<2, 1>()[s] * tensor.template get<3, 2>()[s] -
+          tensor.template get<3, 1>()[s] * tensor.template get<2, 2>()[s];
+      const double tmp4 =
+          tensor.template get<0, 2>()[s] * tensor.template get<1, 3>()[s] -
+          tensor.template get<1, 2>()[s] * tensor.template get<0, 3>()[s];
+      const double tmp5 =
+          tensor.template get<0, 1>()[s] * tensor.template get<1, 3>()[s] -
+          tensor.template get<1, 1>()[s] * tensor.template get<0, 3>()[s];
+      const double tmp6 =
+          tensor.template get<0, 1>()[s] * tensor.template get<1, 2>()[s] -
+          tensor.template get<0, 2>()[s] * tensor.template get<1, 1>()[s];
+
+      det[s] = tensor.template get<0, 0>()[s] *
+                   (tensor.template get<1, 1>()[s] * tmp1 -
+                    tensor.template get<1, 2>()[s] * tmp2 +
+                    tensor.template get<1, 3>()[s] * tmp3) -
+               tensor.template get<1, 0>()[s] *
+                   (tensor.template get<0, 1>()[s] * tmp1 -
+                    tensor.template get<0, 2>()[s] * tmp2 +
+                    tensor.template get<0, 3>()[s] * tmp3) +
+               tensor.template get<2, 0>()[s] *
+                   (tensor.template get<3, 1>()[s] * tmp4 -
+                    tensor.template get<3, 2>()[s] * tmp5 +
+                    tensor.template get<3, 3>()[s] * tmp6) -
+               tensor.template get<3, 0>()[s] *
+                   (tensor.template get<2, 1>()[s] * tmp4 -
+                    tensor.template get<2, 2>()[s] * tmp5 +
+                    tensor.template get<2, 3>()[s] * tmp6);
+    }
+    return det;
+  }
+};
+
+template <typename Index>
+struct DeterminantImpl<Symmetry<1, 1>, Index,
+                       std::enable_if_t<Index::dim == 4>> {
+  template <typename T>
+  static auto apply(const T& tensor) {
+    DataVector det(tensor.template get<0, 0>().size());
```

Comment:
I believe Saul commented that the 4D determinant is terribly slow. We should create an issue and then put a static_assert here referencing the issue.